### PR TITLE
update Sketch menu, add Save hex option

### PR DESCRIPTION
**Pull request by championswimmer**
* Moving Upload options from "File" menu to "Sketch" menu as those
     are sketch actions more than file actions. 

Signed-off-by: Arnav Gupta <championswimmer@gmail.com>

### CODE REVIEW

**Comment by championswimmer** [comment]
Check screenshot for illustration

http://i.imgur.com/z8OXqJ0.png

**Comment by championswimmer** [comment]
Check screenshot for illustration

**Comment by championswimmer** [comment]
@cmaglie This can be considered ?

File: app/src/processing/app/Editor.java [hunk 1]
```diff
@@ -628,14 +615,31 @@ public void actionPerformed(ActionEvent e) {
...
+    item = newJMenuItemShift(_("Upload Using Programmer"), 'U');
+    item.addActionListener(new ActionListener() {
+        public void actionPerformed(ActionEvent e) {
+          handleExport(true);
+        }
+      });
+    sketchMenu.add(item);
+
+
+    item = newJMenuItemAlt("Export compiled Binary", 'S');
```
Comment (cmaglie) [inline app/src/processing/app/Editor.java]: May you add the `_("....")` syntax? so the menu string can be translated.

File: hardware/arduino/avr/platform.txt [hunk 2]
```diff
@@ -67,6 +67,10 @@ recipe.objcopy.eep.pattern="{compiler.path}{compiler.objcopy.cmd}" {compiler.obj
 ## Create hex
 recipe.objcopy.hex.pattern="{compiler.path}{compiler.elf2hex.cmd}" {compiler.elf2hex.flags} {compiler.elf2hex.extra_flags} "{build.path}/{build.project_name}.elf" "{build.path}/{build.project_name}.hex"
 
+## Save hex
+recipe.hex.tmp_file={build.project_name}.hex
```
Comment (cmaglie) [inline hardware/arduino/avr/platform.txt]: Could you rename `hex` here to `output`? and also the related methods `saveHex(..)` to `saveOutput(..)` etc.

I'm wondering that the correct key for this recipe is independent from the output format of the build, so it should really be:

```
recipe.output.tmp_file={build.project_name}.hex
recipe.output.save_file={build.project_name}.{build.variant}.hex
```

**Comment by cmaglie** [comment]
Hi Arnav!

I've added just two simple comments inline. Besides that, it looks ready to be merged.

C

**Comment by championswimmer** [comment]
I see that on sam also we are using the terminology "hex", although the built files are not really "hex" per se.

https://github.com/arduino/Arduino/pull/2567/files#diff-3f519593b982008c8fefa0d4a62300c4L72

That's the reason I used hex in the keys. 
Could you just confirm just once again that I should move on to using recipe.output... instead of recipe.hex...

**Comment by cmaglie** [comment]
Yes, please, let's name the recipe output instead of hex.

You're right about the fact that the inconsistency should be fixed in the objcopy recipe too, it makes no sense to hardcode the .hex extension there. We'll fix that with another pull request BTW.

**Comment by championswimmer** [comment]
@cmaglie done :)

File: arduino-core/src/processing/app/debug/Compiler.java [hunk 3]
```diff
@@ -1042,6 +1051,37 @@ void runRecipe(String recipe) throws RunnerException, PreferencesMapException {
...
+    try {
+      String tmp_file = prefs.getOrExcept("recipe.output.tmp_file");
+      tmp_file = StringReplacer.replaceFromMapping(tmp_file, dict);
+      String save_file = prefs.getOrExcept("recipe.output.save_file");
+      save_file = StringReplacer.replaceFromMapping(save_file, dict);
+
+      File hexFile = new File(prefs.get("build.path") + "/" + tmp_file);
+      File saveFile = new File(sketch.getFolder().getAbsolutePath() + "/" + save_file);
+
+      FileReader in = new FileReader(hexFile);
```
Comment (cmaglie) [inline arduino-core/src/processing/app/debug/Compiler.java]: FileUtils.copyFile may be used here.

File: arduino-core/src/processing/app/debug/Compiler.java [hunk 4]
```diff
@@ -1042,6 +1051,37 @@ void runRecipe(String recipe) throws RunnerException, PreferencesMapException {
...
+      FileWriter out = new FileWriter(saveFile);
+
+      int c;
+      while ((c = in.read()) != -1)
+        out.write(c);
+
+      in.close();
+      out.close();
+
```
Comment (cmaglie) [inline arduino-core/src/processing/app/debug/Compiler.java]: May you add a message for verbose compile? Something like:

```Copying output file xxxxxx.hex to /xxx/yyy/zzz```

Comment (championswimmer) [inline arduino-core/src/processing/app/debug/Compiler.java]: Also for AVR we would like to save the .eep ??

=======================================
  Arnav Gupta (championswimmer)
  LinkedIn <http://in.linkedin.com/in/arnavgupta/> | Web
<http://championswimmer.in> | Blog <http://blog.championswimmer.in> |
Twitter <http://twitter.com/championswimmer> | Facebook
<http://facebook.com/championswimmer> | Github
<http://github.com/championswimmer>

Comment (cmaglie) [inline arduino-core/src/processing/app/debug/Compiler.java]: Yes, maybe we would like to list the files as a comma separated list?

**Comment by ffissore** [comment]
@championswimmer can you update the code so that it merges again?

**Comment by championswimmer** [comment]
Oh shit, had forgotten about this totally. Will do that

=======================================
  Arnav Gupta (championswimmer)
  LinkedIn <http://in.linkedin.com/in/arnavgupta/> | Web
<http://championswimmer.in> | Blog <http://blog.championswimmer.in> |
Twitter <http://twitter.com/championswimmer> | Facebook
<http://facebook.com/championswimmer> | Github
<http://github.com/championswimmer>

**Comment by Nandox7** [comment]
Hey, what's the status on this?
Will the new upload feature allow to upload a saved .hex?

**Pull request closed without merging**